File: product-service/src/main/java/com/ecommerce/productservice/client/member/MemberClientFallbackFactory.java
package com.ecommerce.productservice.client.member;

import com.ecommerce.productservice.client.member.dto.SellerResponse;
import com.ecommerce.productservice.global.exception.ExternalServiceException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cloud.openfeign.FallbackFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;

@Slf4j
@Component
public class MemberClientFallbackFactory implements FallbackFactory<MemberClient> {

    @Override
    public MemberClient create(Throwable cause) {
        log.error("member-service 호출 실패 - {}", cause.getMessage());
        return new MemberClient() {
            @Override
            public ResponseEntity<SellerResponse> getSeller(Long id) {
                log.warn("member-service 불가 - 판매자 정보 없이 응답");
                return ResponseEntity.ok(new SellerResponse(null, null));
            }
        };
    }
}

File: review-service/src/main/kotlin/com/ecommerce/reviewservice/client/member/MemberClientFallbackFactory.kt
package com.ecommerce.reviewservice.client.member

import com.ecommerce.reviewservice.client.member.dto.MemberInfoResponse
import org.slf4j.LoggerFactory
import org.springframework.cloud.openfeign.FallbackFactory
import org.springframework.stereotype.Component

@Component
class MemberClientFallbackFactory : FallbackFactory<MemberClient> {

    private val log = LoggerFactory.getLogger(javaClass)

    override fun create(cause: Throwable?): MemberClient {
        log.error("member-service 호출 실패 - {}", cause?.message)  // ?. 앞의 값이 null이 아니면 뒤으 프로퍼티 호출, null이면 null 반환
        return object : MemberClient {
            override fun getMemberInfos(ids: List<Long>): List<MemberInfoResponse> {
                return emptyList()
            }
        }
    }
}